server/ServerFSM.cpp

```diff
@@ -810,7 +827,9 @@ sc::result MPLobby::react(const AuthResponse& msg) {
     std::string auth;
     ExtractAuthResponseMessageData(message, player_name, auth);
 
-    if (!server.IsAuthSuccess(player_name, auth)) {
+    Networking::AuthRoles roles;
+
+    if (!server.IsAuthSuccess(player_name, auth, roles)) {
         // wrong password
         player_connection->SendMessage(ErrorMessage(str(FlexibleFormat(UserString("ERROR_WRONG_PASSWORD")) % player_name),
```

Comment:
No distinction in the error message between wrong password and right password with a disallowed role?

Reply:
Roles tests in `MPLobby::EstablishPlayer` and returns `ERROR_CLIENT_TYPE_NOT_ALLOWED`. `IsAuthSuccess` only tests password and returns roles without testing them.

Reply:
Don't understand what you mean. `IsAuthSuccess` returns a boolean, not "roles".

Reply:
Should probably return roles rather than fill the passed-in parameter... would be clearer.

Reply:
Need to update the function name then... or have two functions, one of which returns a connection's roles.

Reply:
It is a wrapper for `AuthProvider.is_auth_success`. So it better to rename them.

Reply:
Renamed them.